scheduler/feasible.go

```diff
@@ -200,32 +209,54 @@ func (h *HostVolumeChecker) hasVolumes(n *structs.Node) bool {
 				// state store; this is only possible if the batched fingerprint
 				// update from a delete RPC is written before the delete RPC's
 				// raft entry completes
-				return false
+				return false, FilterConstraintHostVolumes
 			}
 			if vol.State != structs.HostVolumeStateReady {
-				return false
+				return false, FilterConstraintHostVolumes
 			}
 			var capOk bool
 			for _, cap := range vol.RequestedCapabilities {
-				if req.AccessMode == structs.CSIVolumeAccessMode(cap.AccessMode) &&
-					req.AttachmentMode == structs.CSIVolumeAttachmentMode(cap.AttachmentMode) {
+				if req.volumeReq.AccessMode == structs.CSIVolumeAccessMode(cap.AccessMode) &&
+					req.volumeReq.AttachmentMode == structs.CSIVolumeAttachmentMode(cap.AttachmentMode) {
 					capOk = true
 					break
 				}
 			}
 			if !capOk {
-				return false
+				return false, FilterConstraintHostVolumes
 			}
-		} else if !req.ReadOnly {
+
+			if req.volumeReq.Sticky {
+				allocation, err := h.ctx.State().AllocByID(ws, req.allocID)
+				if err != nil {
+					return false, FilterConstraintHostVolumesAllocLookupFailed
+				}
+				if slices.Contains(allocation.HostVolumeIDs, vol.ID) {
+					return true, ""
+				}
+
+				// if an allocation doesn't have a volume ID associated with
+				// it, update it
+				if len(allocation.HostVolumeIDs) == 0 {
+					allocation.HostVolumeIDs = []string{vol.ID}
+					// TODO: figure out how to update allocation. Should we
+					// have a new RPC endpoint for this?
```

Comment:
If an `Allocation` here is a new or updated allocation, then it's getting submitted already to the leader via the `SubmitPlan` RPC.

Reply:
You mean just append this alloc to Plan?

Reply:
Ah, you're right... for some reason I was thinking we had the new `Allocation` already handy but we're calling `Select` on the task group, not the allocation. Maybe trying to add it here doesn't make sense -- can we add it _after_ we've selected the node? (ex. could we write it before we submit the plan, or even as part of plan apply?)

Reply:
Addendum to this: the `computePlacements` function is where we annotate allocations with things like the previous allocation and the reschedule tracker. That suggests a good place to record which volume ID we've given the allocation: we have `Node` and the `Allocation` both handy and we know at that point that they're paired up (because we've _selected_ the node here in `hasVolumes`).